File: packages/mirascript/src/helpers/convert/to-boolean.ts
import type { VmAny } from '../../vm/index.js';
import { VmError } from '../error.js';
import { display } from '../serialize.js';

/** 转换为 boolean */
export function toBoolean<F = undefined>(value: VmAny, fallback?: F): boolean | Exclude<F, undefined> {
    if (typeof value === 'boolean') return value;
    if (fallback === undefined) {
        throw new VmError(`Failed to convert value to boolean: ${display(value)}`, false);
    }
    return fallback as Exclude<F, undefined>;
}

File: crates/python/mirascript/_helpers/convert/to_boolean.py
from __future__ import annotations
from typing_extensions import TypeVar, overload

from ..._vm.types import VmAny
from ..._vm.error import VmError
from ..constants import Uninitialized

T = TypeVar("T")


@overload
def to_boolean(value: VmAny) -> bool: ...
@overload
def to_boolean(value: VmAny, fallback: T) -> bool | T: ...
def to_boolean(value: VmAny, fallback: T = Uninitialized) -> bool | T:
    if isinstance(value, bool):
        return value
    if fallback is Uninitialized:
        raise VmError(f"Cannot convert to boolean: {value}", False)
    return fallback
